File: 2021-06-27/app/src/main/java/com/github/skiiyis/hotfix/HotFixCheckUtil.java
package com.github.skiiyis.hotfix;

import android.content.Context;
import android.util.Log;

import java.io.File;
import java.lang.reflect.Field;

import dalvik.system.DexClassLoader;

public class HotFixCheckUtil {
    public static final String PATCH_NAME = "patch.dex";
    public static final String FIX_CLASS_NAME = "com.github.skiiyis.hotfix.Patch";
    public static final String FIX_MAIN_ACTIVITY = "com.github.skiiyis.hotfix.MainActivity";
    public static final String PATCH_MAIN_ACTIVITY = "com.github.skiiyis.hotfix.MainActivityPatch";

    public static void checkAndHotFix(Context ctx) {
        try {
            DexClassLoader dexClassLoader =
                    new DexClassLoader(new File(ctx.getCacheDir(), PATCH_NAME).getAbsolutePath(), ctx.getCacheDir().getAbsolutePath(), null, ctx.getClass().getClassLoader());
            Class mainActivityClazz = dexClassLoader.loadClass(FIX_MAIN_ACTIVITY);
            Class mainActivityPatchClazz = dexClassLoader.loadClass(PATCH_MAIN_ACTIVITY);
            Field f = mainActivityClazz.getField("__change");
            f.set(null, mainActivityPatchClazz.newInstance());
        } catch (Throwable e) {
            Log.e("HotFix", "Load dexFile error!!", e);
        }
    }
}

File: 2021-06-27/app/src/main/java/com/skiiyis/study/hotfix/HotFixCheckUtil.kt
package com.skiiyis.study.hotfix

import android.content.Context
import android.util.Log
import dalvik.system.DexClassLoader

object HotFixCheckUtil {

    const val FIX_CLASS_NAME = "com.skiiyis.study.hotfix.HotFix"

    // 1. 查找指定的dex是否存在
    // 2. 加载该dex
    // 3. 加载dex中的指定类
    fun checkAndHotFix(ctx: Context) {
        try {
            val dexFilePath = "I'm a path"
            val dexClassLoader =
                DexClassLoader(dexFilePath, ctx.cacheDir.absolutePath, null, ctx.classLoader)
            dexClassLoader.loadClass(FIX_CLASS_NAME)
        } catch (e: Throwable) {
            Log.e("HotFix", "Load dexFile error!!", e)
        }
    }
}